// gatsby-source-datocms/hooks/sourceNodes/createTypes/utils/buildFixedFields.js
"use strict";

var isImage = require('./isImage');

var getSizeAfterTransformations = require('./getSizeAfterTransformations');

var createUrl = require('./createUrl');

var objectAssign = require('object-assign');

module.exports = function () {
  var field = {
    type: 'DatoCmsFixed',
    args: {
      width: {
        type: 'Int',
        defaultValue: 400
      },
      height: 'Int',
      forceBlurhash: 'Boolean',
      imgixParams: 'DatoCmsImgixParams'
    },
    resolve: function resolve(node, _ref) {
      var forceBlurhash = _ref.forceBlurhash,
          width = _ref.width,
          height = _ref.height,
          _ref$imgixParams = _ref.imgixParams,
          imgixParams = _ref$imgixParams === void 0 ? {} : _ref$imgixParams;
      var image = node.entityPayload.attributes;

      if (!isImage(image)) {
        return null;
      }

      if (node.focalPoint && imgixParams.fit === 'crop' && (imgixParams.h || imgixParams.height) && (imgixParams.w || imgixParams.width) && (!imgixParams.crop || imgixParams.crop === 'focalpoint') && !imgixParams['fp-x'] && !imgixParams['fp-y']) {
        imgixParams.crop = 'focalpoint';
        imgixParams['fp-x'] = node.focalPoint.x;
        imgixParams['fp-y'] = node.focalPoint.y;
      }

      var mergedImgixParams = objectAssign({}, imgixParams, {
        w: width
      }, height && {
        h: height
      });

      var _getSizeAfterTransfor = getSizeAfterTransformations(image.width, image.height, mergedImgixParams),
          finalWidth = _getSizeAfterTransfor.width,
          finalHeight = _getSizeAfterTransfor.height;

      var aspectRatio = finalWidth / finalHeight;
      var srcSet = [1, 1.5, 2, 3].map(function (dpr) {
        var extraParams = {
          dpr: dpr
        };

        if (!mergedImgixParams.w && !mergedImgixParams.h) {
          extraParams.w = finalWidth;
        }

        var url = createUrl(image, mergedImgixParams, extraParams, true);
        return "".concat(url, " ").concat(dpr, "x");
      }).join(",\n");
      return {
        aspectRatio: aspectRatio,
        width: finalWidth,
        height: finalHeight,
        format: image.format,
        src: createUrl(image, mergedImgixParams, {}, true),
        srcSet: srcSet,
        forceBlurhash: forceBlurhash
      };
    }
  };
  return {
    fixed: field,
    resolutions: field
  };
};
